src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java

```diff
@@ -377,17 +417,10 @@ public void removeObsoleteServices() throws InterruptedException, ServiceUpdateE
                 }
                 return true;
             }).collect(Collectors.toSet());
-            logger.atInfo(MERGE_CONFIG_EVENT_KEY).kv("service-to-remove", servicesToRemove).log("Removing services");
+
             // waiting for removed service to close before removing reference and config entry
-            for (GreengrassService service : ggServicesToRemove) {
-                try {
-                    service.close().get();
-                } catch (ExecutionException e) {
-                    throw new ServiceUpdateException("Failed to remove obsolete services.", e,
-                            DeploymentErrorCode.REMOVE_COMPONENT_ERROR,
-                            DeploymentErrorCodeUtils.classifyComponentError(service, kernel));
-                }
-            }
+            waitForServicesToClose(ggServicesToRemove, totallyCompleteFuture);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem:**
 This line of code lacks validation when processing input data through the following parameter: &#39;totallyCompleteFuture&#39; (index: 0 | type: CompletableFuture). The parameter is exposed to external callers, because its enclosing class and method are publicly accessible. This means that upstream validation, if it exists, can be bypassed. Other methods validating the parameter: &#39;updateActionForDeployment&#39;, &#39;waitForServicesToStart&#39;, &#39;DeploymentConfigMerger.waitForServicesToClose&#39;. The same parameter type is validated here for example: aws-greengrass-nucleus/src/main/java/com/aws/greengrass/deployment/DeploymentConfigMerger.java:123. Malicious, malformed, or unbounded inputs can cause unexpected runtime behavior or crashes, and can slow performance.

**Fix:**
 Add checks to ensure the validity of the parameter&#39;s value, such as testing it for nullness (for example, using the @NonNull annotation as [described in the Lombok library](https://projectlombok.org/features/NonNull)), emptiness, or equality. Or to prevent direct calls to it, [reduce the method&#39;s visibility](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html).

 Learn more about potential threats and guidance from the [Common Weakness Enumeration website](https://cwe.mitre.org/data/definitions/693.html) and the [OWASP Cheat Sheet series](https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).